service/test/BatchServerTest.cpp

```diff
@@ -395,77 +352,78 @@ TEST_F(BatchServerTest, run_batch_write_empty)
 
-    m_batch_server->create_shmem();
+    m_batch_server_pid->create_shmem();
 
     struct stat data;
     int result = stat((shmem_prefix_signal).c_str(), &data);
     ASSERT_EQ(0, result);                                // check that the file exists
-    EXPECT_EQ((size_t)data.st_size, signal_shmem_size);  // check the size of the file
-    EXPECT_EQ(data.st_uid, uid);                         // check the chown succeeded in uid
-    EXPECT_EQ(data.st_gid, gid);                         // check the chown succeeded in gid
+    EXPECT_EQ(signal_shmem_size, (size_t)data.st_size);  // check the size of the file
+    EXPECT_EQ(uid, data.st_uid);                         // check the chown succeeded in uid
+    EXPECT_EQ(gid, data.st_gid);                         // check the chown succeeded in gid
 
     result = stat(shmem_prefix_control.c_str(), &data);
     ASSERT_EQ(0, result);                                 // check that the file exists
-    EXPECT_EQ((size_t)data.st_size, control_shmem_size);  // check the size of the file
-    EXPECT_EQ(data.st_uid, uid);                          // check the chown succeeded in uid
-    EXPECT_EQ(data.st_gid, gid);                          // check the chown succeeded in gid
+    EXPECT_EQ(control_shmem_size, (size_t)data.st_size);  // check the size of the file
+    EXPECT_EQ(uid, data.st_uid);                          // check the chown succeeded in uid
+    EXPECT_EQ(gid, data.st_gid);                          // check the chown succeeded in gid
 
     EXPECT_EQ(0, unlink(shmem_prefix_signal.c_str()));
     EXPECT_EQ(0, unlink(shmem_prefix_control.c_str()));
 }
 
 TEST_F(BatchServerTest, fork_with_setup)
 {
-    SetUp_full_server();
-
     size_t *counter_mem = (size_t*) mmap(
         NULL,                        // void *addr
         sizeof(size_t),              // size_t length
         PROT_READ  | PROT_WRITE,     // int prot
         MAP_SHARED | MAP_ANONYMOUS,  // int flags
         -1,                          // int fd
-        0                            // off_t offset
-    );
+        0);                          // off_t offset
     size_t &counter = *counter_mem;
     counter = 0;
 
     std::function<void(void)> setup = [&counter](void)
     {
-        ASSERT_EQ(0u, counter);
-        ++counter;
+        if (counter == 0u) {
+            ++counter;
+        }
     };
 
-    std::function<void(void)> run = [&counter](void)
+    std::function<void(void)> run = [&counter, this](void)
     {
-        ASSERT_EQ(1u, counter);
-        ++counter;
+        if (counter == 1u) {
+            ++counter;
+        }
+        this->TearDown();
+        testing::Mock::AllowLeak(m_signal_shmem.get());
+        testing::Mock::AllowLeak(m_control_shmem.get());
+        testing::Mock::AllowLeak(m_pio_ptr.get());
+        testing::Mock::AllowLeak(m_batch_status.get());
+        testing::Mock::AllowLeak(m_posix_signal.get());
```

Comment:
Can we do something like
```cpp
this->m_signal_shmem.reset();
```

Reply:
If there are any copies of the shared pointer elsewhere, the reset() will not be sufficient.  In this case we know that the _Exit() call will not clean up these mocks, and so we can use the gmock feature.

Reply:
m_signal_shmem and m_control_shmem should *not* be leaked since they are managed by the forked process explicitly.  These two lines should be removed.